# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

In [9]:
text[:82]

'jerry: do you know what this is all about? do you know, why were here? to be out, '

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [10]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [11]:
# Check the type of the text
type(text)

str

In [12]:
import problem_unittests as tests

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    
    vocab_to_int = {}
    
    #print(type(text))
    #print(text)
    i=0
    for word in text:
        if word not in vocab_to_int.keys():
            vocab_to_int[word] = i
            i += 1 
     
    
    # print(vocab_to_int)
    
    int_to_vocab= {key: word for word, key in vocab_to_int.items()}
    
    # return tuple
    return (vocab_to_int, int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [13]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function

  
    lookup_dict = {".": "||Period||",
                  ",": "||Comma||",
                  '"':"||Quotation_Mark||",
                  ';':"||Semicolon||",
                   "!":"||Exclamation_mark||",
                   "?":"||Question_mark||",
                   "(": "||Left_Parentheses||",
                   ")": "||Right_Parentheses||",
                   "-": "||Dash||",
                   "\n": "||Return||"
                  }
    
    return lookup_dict

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to look at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [14]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

In [16]:
# Check the types of the data you are working with
type(int_text), type(vocab_to_int), type(int_to_vocab), type(token_dict)

(list, dict, dict, dict)

In [17]:
# Print the 10 values of int_text
int_text[:10]

[0, 1, 2, 3, 3, 3, 4, 2, 1, 5]

In [18]:
# Print 10 items from vocab_to_int
for ex in list(vocab_to_int.items())[:10]:
    print(ex)

('this', 0)
('is', 1)
('out', 2)
('||period||', 3)
('and', 4)
('one', 5)
('of', 6)
('the', 7)
('single', 8)
('most', 9)


In [19]:
# Print 10 items from int_to_vocab
for ex in list(int_to_vocab.items())[:10]:
    print(ex)

(0, 'this')
(1, 'is')
(2, 'out')
(3, '||period||')
(4, 'and')
(5, 'one')
(6, 'of')
(7, 'the')
(8, 'single')
(9, 'most')


In [20]:
# Print 10 items from token_dict
for ex in list(token_dict.items())[:10]:
    print(ex)

('.', '||Period||')
(',', '||Comma||')
('"', '||Quotation_Mark||')
(';', '||Semicolon||')
('!', '||Exclamation_mark||')
('?', '||Question_mark||')
('(', '||Left_Parentheses||')
(')', '||Right_Parentheses||')
('-', '||Dash||')
('\n', '||Return||')


## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [21]:
## Check some the Cuda version
import torch

print(torch.version.cuda)

11.1


In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')
else:
    print("Congrats, you have GPU")

Congrats, you have GPU


## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [28]:
from torch.utils.data import TensorDataset, DataLoader


def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    
    feature_tensors, target_tensors = list(), list()
    
    len_words = len(words)
    
    for i in range(len_words):
        idx_target = i + sequence_length
        if idx_target < len_words:
            feature_tensors.append(words[i:idx_target])
            target_tensors.append(words[idx_target])
            
    data = TensorDataset(torch.from_numpy(np.array(feature_tensors)),
                         torch.from_numpy(np.array(target_tensors)))
    data_loader = DataLoader(data, shuffle=True, batch_size=batch_size)
    return data_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own

data = iter(batch_data(range(50), 10,4))
print(next(data))
print(next(data))

[tensor([[33, 34, 35, 36, 37, 38, 39, 40, 41, 42],
        [ 9, 10, 11, 12, 13, 14, 15, 16, 17, 18],
        [38, 39, 40, 41, 42, 43, 44, 45, 46, 47],
        [35, 36, 37, 38, 39, 40, 41, 42, 43, 44]]), tensor([43, 19, 48, 45])]
[tensor([[25, 26, 27, 28, 29, 30, 31, 32, 33, 34],
        [23, 24, 25, 26, 27, 28, 29, 30, 31, 32],
        [24, 25, 26, 27, 28, 29, 30, 31, 32, 33],
        [ 7,  8,  9, 10, 11, 12, 13, 14, 15, 16]]), tensor([35, 33, 34, 17])]


### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [29]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[36, 37, 38, 39, 40],
        [14, 15, 16, 17, 18],
        [15, 16, 17, 18, 19],
        [35, 36, 37, 38, 39],
        [ 3,  4,  5,  6,  7],
        [41, 42, 43, 44, 45],
        [40, 41, 42, 43, 44],
        [42, 43, 44, 45, 46],
        [21, 22, 23, 24, 25],
        [44, 45, 46, 47, 48]])

torch.Size([10])
tensor([41, 19, 20, 40,  8, 46, 45, 47, 26, 49])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [30]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # TODO: Implement function

    
        # set class variables
        self.output_size = output_size 
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim
        
        # define model layers
        
        self.embed = nn.Embedding(vocab_size,embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers, dropout= dropout,batch_first= True)
        self.fc = nn.Linear(hidden_dim, output_size)
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function 
        
        # Get the batch size of the input
        batch_size = nn_input.size(0)
        
        # Get the vector representation of the input

            
        embed = self.embed(nn_input.long())
        
        # Put the input & hidden to the lstm layer to get the output
        
        output_lstm, hidden_output= self.lstm(embed, hidden)
        
    
        output_lstm = output_lstm.contiguous().view(-1, self.hidden_dim)
        
        # Get the scores of the word
        output = self.fc(output_lstm)
        
        # reshape into (batch_size, seq_length, output_size)
        output = output.view(batch_size, -1, self.output_size)
        
        # get last batch
        out = output[:, -1]
        
        # return one batch of output word scores and the hidden state
        return out, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        
        weight = next(self.parameters()).data
        
        if (train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())
        
        return hidden


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [31]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param rnn: The PyTorch Module that holds the neural network
    :param optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    
    # move data to GPU, if available
    
    if train_on_gpu:
        inp, target = inp.cuda(), target.cuda()
        
    hidd = tuple([each.data for each in hidden])
    
    rnn.zero_grad()
    
    output, hidd = rnn(inp, hidd)
    
    # calculate the loss and perform backpropagation
        
    loss = criterion(output, target.long())
    
    # Computes the gradient of current tensor w.r.t. graph leaves.
    loss.backward()
    
    # perform backpropagation and optimization
    nn.utils.clip_grad_norm_(rnn.parameters(), 5)
    optimizer.step()
    
    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidd

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [32]:
import time

In [33]:
import time
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        start = time.time()
        epoch_start  = time.time()
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            
            if batch_i % show_every_n_batches == 0:
                message = 'Epoch: {:>4}/{:<4} Batch: {:>4}/{:<4}  Loss: {:.4} ; Took: {} seconds \n'.format(
                    epoch_i, n_epochs, 
                    batch_i, n_batches,
                    np.average(batch_losses), int(time.time()-start ))
                print(message)

                start = time.time()
                batch_losses = []

        print('Epoch {} took {} seconds in total'.format(epoch_i, int(time.time()-epoch_start)))
        print('===========\n')


    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of unique tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [38]:
# Data params
# Sequence Length
sequence_length = 12  # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [40]:
# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 500
# Hidden Dimension
hidden_dim = 600
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 500

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [41]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 10 epoch(s)...
Epoch:    1/10   Batch:  500/13939  Loss: 5.799 ; Took: 9 seconds 

Epoch:    1/10   Batch: 1000/13939  Loss: 5.063 ; Took: 9 seconds 

Epoch:    1/10   Batch: 1500/13939  Loss: 4.931 ; Took: 9 seconds 

Epoch:    1/10   Batch: 2000/13939  Loss: 4.818 ; Took: 9 seconds 

Epoch:    1/10   Batch: 2500/13939  Loss: 4.677 ; Took: 9 seconds 

Epoch:    1/10   Batch: 3000/13939  Loss: 4.64 ; Took: 9 seconds 

Epoch:    1/10   Batch: 3500/13939  Loss: 4.58 ; Took: 9 seconds 

Epoch:    1/10   Batch: 4000/13939  Loss: 4.511 ; Took: 9 seconds 

Epoch:    1/10   Batch: 4500/13939  Loss: 4.504 ; Took: 9 seconds 

Epoch:    1/10   Batch: 5000/13939  Loss: 4.438 ; Took: 9 seconds 

Epoch:    1/10   Batch: 5500/13939  Loss: 4.443 ; Took: 9 seconds 

Epoch:    1/10   Batch: 6000/13939  Loss: 4.391 ; Took: 9 seconds 

Epoch:    1/10   Batch: 6500/13939  Loss: 4.369 ; Took: 9 seconds 

Epoch:    1/10   Batch: 7000/13939  Loss: 4.325 ; Took: 9 seconds 

Epoch:    1/10   Batch

Epoch:    5/10   Batch: 5000/13939  Loss: 3.606 ; Took: 9 seconds 

Epoch:    5/10   Batch: 5500/13939  Loss: 3.614 ; Took: 9 seconds 

Epoch:    5/10   Batch: 6000/13939  Loss: 3.613 ; Took: 9 seconds 

Epoch:    5/10   Batch: 6500/13939  Loss: 3.617 ; Took: 9 seconds 

Epoch:    5/10   Batch: 7000/13939  Loss: 3.617 ; Took: 9 seconds 

Epoch:    5/10   Batch: 7500/13939  Loss: 3.636 ; Took: 9 seconds 

Epoch:    5/10   Batch: 8000/13939  Loss: 3.614 ; Took: 9 seconds 

Epoch:    5/10   Batch: 8500/13939  Loss: 3.636 ; Took: 9 seconds 

Epoch:    5/10   Batch: 9000/13939  Loss: 3.683 ; Took: 9 seconds 

Epoch:    5/10   Batch: 9500/13939  Loss: 3.645 ; Took: 9 seconds 

Epoch:    5/10   Batch: 10000/13939  Loss: 3.643 ; Took: 9 seconds 

Epoch:    5/10   Batch: 10500/13939  Loss: 3.629 ; Took: 9 seconds 

Epoch:    5/10   Batch: 11000/13939  Loss: 3.648 ; Took: 9 seconds 

Epoch:    5/10   Batch: 11500/13939  Loss: 3.643 ; Took: 9 seconds 

Epoch:    5/10   Batch: 12000/13939  Loss: 3

Epoch:    9/10   Batch: 10000/13939  Loss: 3.395 ; Took: 9 seconds 

Epoch:    9/10   Batch: 10500/13939  Loss: 3.412 ; Took: 9 seconds 

Epoch:    9/10   Batch: 11000/13939  Loss: 3.427 ; Took: 9 seconds 

Epoch:    9/10   Batch: 11500/13939  Loss: 3.425 ; Took: 9 seconds 

Epoch:    9/10   Batch: 12000/13939  Loss: 3.414 ; Took: 9 seconds 

Epoch:    9/10   Batch: 12500/13939  Loss: 3.451 ; Took: 9 seconds 

Epoch:    9/10   Batch: 13000/13939  Loss: 3.477 ; Took: 9 seconds 

Epoch:    9/10   Batch: 13500/13939  Loss: 3.475 ; Took: 9 seconds 

Epoch 9 took 272 seconds in total

Epoch:   10/10   Batch:  500/13939  Loss: 3.331 ; Took: 11 seconds 

Epoch:   10/10   Batch: 1000/13939  Loss: 3.204 ; Took: 11 seconds 

Epoch:   10/10   Batch: 1500/13939  Loss: 3.27 ; Took: 9 seconds 

Epoch:   10/10   Batch: 2000/13939  Loss: 3.217 ; Took: 10 seconds 

Epoch:   10/10   Batch: 2500/13939  Loss: 3.245 ; Took: 10 seconds 

Epoch:   10/10   Batch: 3000/13939  Loss: 3.254 ; Took: 10 seconds 

E

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [46]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [47]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    # Return a new array of given shape and type, filled with fill_value.
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities        
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        if(train_on_gpu):
            current_seq = current_seq.cpu() # move to cpu
        # the generated word becomes the next "current sequence" and the cycle can continue
        # import pdb; pdb.set_trace()
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [48]:
# run the cell multiple times to get different results!
gen_length = 300 # modify the length to your preference
prime_word = 'elaine' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

elaine: announcer frog mouths opened up the window.

george:(pleading) what?

jerry: well, i don't want to hear it.

kramer:(indicating) well, i got it. i was just wondering what i'm gonna tell you, i can't get it out of here.

kramer: well, i don't know how you feel.

jerry: you don't know, i can't. but i don't think we are gonna do it. i don't have any furniture.

elaine: i know, but i can't help her.

elaine: i don't want to talk to him, but...

jerry: oh yeah...

george: you know, i've been thinking about it, jerry. i just came in. i got to go out with him.

jerry:(to the phone) yeah.

kramer: hey, hey! you know i got a hopper?

jerry: yeah, yeah.

elaine: what is that about?

jerry: what? why?

george: well, you know the point of the post office. the spit of the aryan union. the horror, they have no choice. the only way you know, i think he's in the hospital. i just wanted you to know me.

elaine:(leaving) i know, but it's a good thing.(jerry picks it up)

george: so, what do you 

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [49]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.